### DISCUSSION TITLE
테스트코드 작성 시 성공 케이스와 실패 케이스 중 무엇을 주요로 작성하시고 그 이유는 무엇이신가요?

### DISCUSSION THREAD

**Discussion by wonsss** [woowacourse-study/2022-code-review-study#20] [Lotto] [answered]
저는 성공 케이스만 확인하는 테스트를 주로 작성했는데, 이것만으로는 반쪽짜리 테스트라는 생각이 들었어요. 리뷰어님께서도 안되는 케이스를 주요로 작성할 것을 권하셨는데요.

여러분들은 테스트코드 작성 시 성공 케이스와 실패 케이스 중 무엇을 주요로 작성하시고 그 이유는 무엇이신지 궁금합니다. 😄

**Comment by lokba** [comment]
제가 오늘 리뷰정리를 하면서, 알게된 좋은 말입니다.
해당 discussion과 관련이 있다고 생각하여 아래에 글 남깁니다.

> 사용자 편의성에 대한 좋은 조언
> 먼저 사용자가 무엇을 해야할지 직관적으로 바로 알 수 없다면 개발자의 잘못이라고 생각합니다
>  직관적으로 알 수 있는 UI가 먼저이고 그럴 수 없는 경우라면 어떻게 해야할지 구체적이고 상세하게 안내해야 된다고 생각합니다 그리고 악의적인 초등학생이 깨트리려고 해도 절대 깨지지 않는 앱을 만들어야 된다고 생각합니다(이건 Bsidesoft의 맹대표님 강의 중에 들은 교훈입니다)
>    그래서 성공하는 일부 케이스만 테스트하는게 아니라 에러를 더 많이 발생시키고 그에 대한 예외처리를 상세히 기술해서 사용성이 깨지는 경우를 방지하려고 노력합니다


이 말을 듣고보니,
저는 실패 케이스를 주요로 작성해야 할 것 같습니다.

**Comment by wonsss** [comment, marked as answer]
저도 발리스타님께서 이와 관련하여 다음과 같은 리뷰를 주셨었는데, 내용이 너무 좋아서 공유드리고 싶네요. 😄 

> 로직의 "정상적으로 작동" 하는 결과는 하나뿐입니다. 하지만 오작동하는 케이스는 수십가지가 넘죠. 예를들자면 결과값이 "2"가 되어야 하는 케이스가 있다고 합시다. 결국 어떤 시나리오, 2+0, 1+1, 0+2 등, 결국 2가 되면 맞는 게임입니다. 하지만 그 외에 아닌 케이스는 정말 많겠죠?
서비스의 개발도 위와 동일합니다. 결국 어떻게 되었던 이 로직이 제대로 도는 결과는 하나이고 오작동하는 케이스는 수십가지가 넘을 것입니다. 특히 복잡한 유저관점에서 해석을 한다면 유저가 어떻게 행동할지 미리 생각해서 유저가 어떤 행동을 했을때 오작동을 하면 "유저에게 이렇게 사용하지마!" 라는 피드백이 필요할 것입니다.
이는 기능관점에서도 유효합니다. "버튼을 누른다"의 경우 정상적인 케이스는 "버튼이 눌려서 어떤 액션이 실행되었는가?" 겠죠. 그런데 버튼을 0.1초 안에 200번을 누르는 케이스가 존재한다면 (실제로 dom 이벤트를 이렇게 증식 시켜서 디도스를 할 수도 있겠죠.) 이런 문제 케이스에 대해서 대응을 해야겠죠? 왜냐하면 버튼에서 "어떤 액션"의 범주는 굉장히 넓고, 거기엔 비동기 처리도 있을테니까요.
결론적으로 안되는 케이스를 주요로 작성하라는건 다음과 같습니다.
애자일 철학에서도 나오는 "유저스토리 관점으로 생각하라" 처럼, 이 메소드/함수와 같은 행동을 사용하는 유저 관점에서 생각하고, 테스트를 작성하라
유저 관점에서 생각하면 유저가 정상적으로 사용할 수 있는 케이스는 "하나" 이고, 그 외에 오작동 하는 케이스에 대한 "수많은" 문제를 정상적으로 사용할 수 있도록 유도하는 코드가 등장하게 된다.
테스트 코드는 협업하는 사람들에게 어떤 문제에 대해 코드가 보완되어 있는지 공유하는 성격도 갖고있습니다. 특히 현업에선 QA TC등이 결국 테스트 코드와 매칭되는 부분이 많습니다. 에러 상황에 대한 테스트 코드가 많을수록 당연히 QA에서 커버되는 커버리지를 코드단위로 이행시켜서 효율성을 증대 시킬 수 있겠죠?
테스트 코드로 1박 2일도 떠들수 있는데 여기서 그만하도록 하겠습니다.. 대충 이래요. ㅋㅋㅋㅋ :)
테스트 코드는 중요하고, 서비스에 영향도 크고, 서비스를 벗어나 개발, QA 등 다양한 사람들에게 선한 영향력을 줄 수 있습니다.

조금 더 정리하자면, 

- 로직의 정상 작동 결과는 하나뿐이지만, 오작동하는 케이스는 수십가지가 넘는다.
	- 따라서 유저가 어떤 행동을 했을 때 오작동을 하면 유저에게 그에 대한 피드백이 필요하다. 이러한 오작동 케이스에 대한 수많은 문제를 정상적으로 사용할 수 있도록 유도하는 코드가 필요하다.
- 테스트코드는 오작동 문제 상황에 대한 코드 보완 여부를 협업하는 사람들과 공유하는 성격도 갖고 있다. 
	- 따라서 에러 상황에 대한 테스트코드를 많이 고민하고 작성수록, 이 테스트코드를 코드단위로 이행시켜서 효율성을 증대시킬 수 있다.
- 결론: 유저 관점에서 유저가 일으킬 수 있는 수많은 오작동 케이스를 커버할 수 있는 테스트코드를 작성하자. 그리고 테스트코드는 에러 상황 커버리지를 코드단위로 이행시키며 다양한 사람들에게 선한 영향력을 줄 수 있음을 염두하고 정성껏 작성하자!

**Reply by lokba** [reply]
좋은 말 감사합니다 :)
